pkg/controllers/node/startup.go

```diff
@@ -0,0 +1,74 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/utils/injectabletime"
+	"github.com/aws/karpenter/pkg/utils/node"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const StartupTimeout = 15 * time.Minute
+
+// Startup is a subreconciler that
+// 1. Removes the NotReady taint when the node is ready. This taint is originally applied on node creation.
+// 2. Terminates nodes that don't transition to ready within StartupTimeout
+type Startup struct {
+	kubeClient client.Client
+}
+
+// Reconcile reconciles the node
+func (r *Startup) Reconcile(ctx context.Context, _ *v1alpha5.Provisioner, n *v1.Node) (reconcile.Result, error) {
+	if !notReadyTaintKeyExists(n) {
+		// At this point, the startup of the node is complete and no more evaluation is necessary.
+		return reconcile.Result{}, nil
+	}
+
+	if !node.IsReady(n) {
+		if timeSinceCreation := injectabletime.Now().Sub(n.GetCreationTimestamp().Time); timeSinceCreation < StartupTimeout {
+			return reconcile.Result{RequeueAfter: StartupTimeout - timeSinceCreation}, nil
+		}
+		logging.FromContext(ctx).Infof("Triggering termination for node that failed to transition to ready")
+		if err := r.kubeClient.Delete(ctx, n); err != nil {
+			return reconcile.Result{}, fmt.Errorf("deleting node, %w", err)
+		}
+		return reconcile.Result{}, nil
+	}
+	taints := []v1.Taint{}
+	for _, taint := range n.Spec.Taints {
+		if taint.Key != v1alpha5.NotReadyTaintKey {
+			taints = append(taints, taint)
+		}
+	}
+	n.Spec.Taints = taints
+	return reconcile.Result{}, nil
+}
+
+func notReadyTaintKeyExists(n *v1.Node) bool {
```

Comment:
Thoughts on making a similar helper to `taints.Has`. Perhaps `taints.HasKey(v1alpha5.NotReadyTaintKey)`

Reply:
Done, thanks for the suggestion!